SecondSemester/ParseTree/ParseTree.cs

```diff
@@ -0,0 +1,199 @@
+        if (this.root == null)
+        {
+            throw new IncorrectInputException();
+        }
+
+        return this.root.Calculate();
+    }
+
+    /// <summary>
+    /// Returns the string representation of the parse tree.
+    /// </summary>
+    /// <returns>The string representation of the parse tree.</returns>
+    public override string ToString() => this.root == null ? "\n" : this.root.ToString();
+
+    private bool IsOperator(string token)
+    {
+        switch (token)
+        {
+            case "+":
+            case "-":
+            case "*":
+            case "/":
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    private Operator? BuildTree(string stringTree)
+    {
+        var stringTreeWithoutParentheses = stringTree.Replace("(", string.Empty);
+        stringTreeWithoutParentheses = stringTreeWithoutParentheses.Replace(")", string.Empty);
+
+        var elements = stringTreeWithoutParentheses.Split(' ');
+        var splitFromOriginal = stringTree.Split(' ');
+        var subtrees = new Stack<Operator>();
+
+        Operator? currentSubtree = null;
+        for (int i = 0; i < elements.Length; ++i)
+        {
+            var element = elements[i];
+
+            if (subtrees.Count == 0 && currentSubtree is { Right: not null })
+            {
+                throw new IncorrectInputException();
+            }
+
+            if (this.IsOperator(element))
+            {
+                if (splitFromOriginal[i] != $"({element}")
+                {
+                    throw new IncorrectInputException();
+                }
+
+                var newSubtree = new Operator(element);
+                if (currentSubtree is not null)
+                {
+                    subtrees.Push(currentSubtree);
+                }
+
+                currentSubtree = newSubtree;
+            }
+            else
+            {
+                if (currentSubtree is null)
+                {
+                    throw new IncorrectInputException();
+                }
+
+                var operand = new Operand(element);
+                currentSubtree.Update(operand);
+
+                var parentheses = string.Empty;
+                while (currentSubtree.Right is not null && subtrees.TryPop(out var oldSubtree))
+                {
+                    parentheses += ')';
+                    oldSubtree.Update(currentSubtree);
+                    currentSubtree = oldSubtree;
+                }
+
+                if (currentSubtree.Right != null)
+                {
+                    parentheses += ')';
+                }
+
+                if (splitFromOriginal[i] != $"{element}{parentheses}")
+                {
+                    throw new IncorrectInputException();
+                }
+            }
+        }
+
+        return currentSubtree;
+    }
+
+    private abstract class Node(string value)
+    {
+        protected string Value { get; private set; } = value;

```

Comment:
```suggestion
        protected string Value { get; init; } = value;
```
Почему protected, кстати?